kuzzle-sdk-java/src/main/java/io/kuzzle/sdk/Kuzzle.java

```diff
@@ -0,0 +1,284 @@
+    }
+
+    /** Initialize a new instance of Kuzzle
+     * @param networkProtocol   The network protocol
+     * @param options           Kuzzle options
+     * @throws IllegalArgumentException
+     */
+    public Kuzzle(
+            AbstractProtocol networkProtocol,
+            final KuzzleOptions options
+    ) throws IllegalArgumentException {
+
+        if (networkProtocol == null) {
+            throw new IllegalArgumentException("networkProtocol can't be null");
+        }
+
+        KuzzleOptions kOptions = options != null
+                ? options
+                : new KuzzleOptions();
+
+        this.networkProtocol = networkProtocol;
+        this.networkProtocol.registerResponseEvent(this::onResponseReceived);
+        this.networkProtocol.registerStateChangeEvent(this::onStateChanged);
+
+        this.maxQueueSize = new AtomicInteger(kOptions.getMaxQueueSize());
+        this.minTokenDuration = new AtomicInteger(kOptions.getMinTokenDuration());
+        this.refreshedTokenDuration = new AtomicInteger(kOptions.getRefreshedTokenDuration());
+        this.maxRequestDelay = new AtomicInteger(kOptions.getMaxRequestDelay());
+
+        this.version = "3.0.0";
+        this.instanceId = UUID.randomUUID().toString();
+
+        this.tokenExpiredEvent = new EventListener();
+        this.unhandledResponseEvent = new EventListener<>();
+    }
+
+    /** Establish a network connection
+     * @throws Exception
+     */
+    public void connect() throws Exception {
+        networkProtocol.connect();
+    }
+
+    /**
+     * Disconnect this instance
+     */
+    public void disconnect() {
+        networkProtocol.disconnect();
+    }
+
+    /** Handles the ResponseReceivedEvent from the network protocol
+     * @param payload Raw API Response
+     */
+    protected void onResponseReceived(String payload) {
+
+        Response response = new Response();
+        response.fromMap(JsonSerializer.deserialize(payload));
+
+        if (response.room != null
+            && requests.containsKey(notNull(response.room, ""))
+        ) {
+            if (response.error != null) {
+                if (response.error.message != null
+                    && response.error.message.equals("Token expired")
+                ) {
+                    tokenExpiredEvent.trigger();
+                }
+
+                Task<Response> task = requests.get(
+                        notNull(response.requestId, ""));
+                if (task != null) {
+                    task.setException(new ApiErrorException(response));
+                }
+
+            } else {
+                Task<Response> task = requests.get(
+                        notNull(response.requestId, ""));
+
+                if (task != null) {
+                    task.trigger(response);
+                }
+
+                requests.remove(notNull(response.requestId, ""));
+            }
+        } else {
+            unhandledResponseEvent.trigger(response);
```

Comment:
Nitpicking: it's better to use a guard and return instead of indenting the method code